File: backend/app/routes/class_requests.py
"""班级变更审批路由。"""
from flask import Blueprint, request
from flask_jwt_extended import get_jwt_identity, jwt_required

from app.services.class_request import ClassRequestService
from app.utils.access import is_admin
from app.utils.decorators import role_required
from app.utils.response import error_response, success_response

class_requests_bp = Blueprint('class_requests', __name__, url_prefix='/api/v1/class-requests')


@class_requests_bp.route('', methods=['GET'])
@jwt_required()
@role_required('teacher', 'admin')
def list_class_requests():
    try:
        from app.models import User

        user_id = int(get_jwt_identity())
        user = User.query.get(user_id)
        status = request.args.get('status')
        if is_admin(user):
            rows = ClassRequestService.list_requests(status=status)
        else:
            rows = ClassRequestService.list_requests(status=status, requester_id=user_id)
        return success_response(rows)
    except Exception as exc:
        return error_response(str(exc), 50001, None, 500)


@class_requests_bp.route('', methods=['POST'])
@jwt_required()
@role_required('teacher')
def submit_class_request():
    try:
        user_id = int(get_jwt_identity())
        data = request.get_json() or {}
        action = data.get('action')
        class_id = data.get('class_id', type=int) if data.get('class_id') is not None else None
        payload = data.get('payload') or {}
        reason = data.get('reason')
        row = ClassRequestService.create_request(user_id, action, payload, class_id=class_id, reason=reason)
        return success_response(row.to_dict(), '申请已提交，等待管理员审核', 0, 201)
    except PermissionError as exc:
        return error_response(str(exc), 40301, None, 403)
    except ValueError as exc:
        return error_response(str(exc), 40001, None, 400)
    except Exception as exc:
        return error_response(str(exc), 50001, None, 500)


@class_requests_bp.route('/<int:request_id>/review', methods=['POST'])
@jwt_required()
@role_required('admin')
def review_class_request(request_id):
    try:
        reviewer_id = int(get_jwt_identity())
        data = request.get_json() or {}
        approve = bool(data.get('approve'))
        note = data.get('note')
        row = ClassRequestService.review(request_id, reviewer_id, approve, note)
        label = '已通过' if row.status == 'approved' else '已驳回'
        return success_response(row.to_dict(), f'申请{label}')
    except PermissionError as exc:
        return error_response(str(exc), 40301, None, 403)
    except ValueError as exc:
        return error_response(str(exc), 40001, None, 400)
    except Exception as exc:
        return error_response(str(exc), 50001, None, 500)

File: frontend/src/api/classRequests.ts
import { http, type ApiEnvelope } from './http'

export type ClassRequestAction = 'create' | 'update' | 'delete'
export type ClassRequestStatus = 'pending' | 'approved' | 'rejected'

export interface ClassChangeRequest {
  id: number
  requester_id: number
  requester_name: string | null
  action: ClassRequestAction
  status: ClassRequestStatus
  class_id: number | null
  class_name: string | null
  payload: Record<string, unknown>
  reason: string | null
  reviewer_id: number | null
  review_note: string | null
  created_at: string | null
  reviewed_at: string | null
}

export async function fetchClassRequests(status?: ClassRequestStatus) {
  const { data } = await http.get<ApiEnvelope<ClassChangeRequest[]>>('/v1/class-requests', {
    params: status ? { status } : undefined,
  })
  if (data.code !== 0) throw new Error(data.message || '加载申请失败')
  return data.data
}

export async function submitClassRequest(body: {
  action: ClassRequestAction
  class_id?: number
  payload: Record<string, unknown>
  reason?: string
}) {
  const { data } = await http.post<ApiEnvelope<ClassChangeRequest>>('/v1/class-requests', body)
  if (data.code !== 0) throw new Error(data.message || '提交申请失败')
  return data.data
}

export async function reviewClassRequest(requestId: number, approve: boolean, note?: string) {
  const { data } = await http.post<ApiEnvelope<ClassChangeRequest>>(
    `/v1/class-requests/${requestId}/review`,
    { approve, note },
  )
  if (data.code !== 0) throw new Error(data.message || '审批失败')
  return data.data
}
